File: packages/perception/src/scheduler.ts
/**
 * Keeps one large model in memory at a time.
 *
 * A 16 GB card is the declared target, and speech recognition, a vision
 * embedder and a vision-language model will not co-exist in it. The pipeline is
 * already staged — transcribe everything, then embed everything, then describe
 * the events that need it — so the fix is not smaller models but never asking
 * for two at once.
 *
 * The scheduler makes that a rule instead of a convention: work that needs
 * accelerator memory takes a lease, and a lease for a different model waits for
 * the previous one to be released. Work in the same slot runs back to back
 * without reloading, which is where nearly all of the time goes.
 */
export interface ModelLease {
  /** Slot name, typically the capability: `speech`, `visual`, `context`. */
  readonly slot: string;
  release(): void;
}

export interface ModelSchedulerOptions {
  /** How many slots may be resident at once. One is the safe default for 16 GB. */
  concurrency?: number;
  /** Called when a slot is evicted, so a backend can unload. */
  onEvict?: (slot: string) => void | Promise<void>;
}

export class ModelScheduler {
  private readonly concurrency: number;
  private readonly onEvict: ((slot: string) => void | Promise<void>) | undefined;
  private resident: string[] = [];
  private activeLeases = 0;
  private readonly queue: { slot: string; grant: () => void }[] = [];

  constructor(options: ModelSchedulerOptions = {}) {
    this.concurrency = Math.max(1, options.concurrency ?? 1);
    this.onEvict = options.onEvict;
  }

  /** Slots currently considered loaded, most recently used last. */
  get residentSlots(): readonly string[] {
    return this.resident;
  }

  /**
   * Runs `work` while holding a lease on `slot`.
   *
   * Consecutive calls for the same slot never evict anything, which is what
   * makes "transcribe all thirty files" cost one model load instead of thirty.
   */
  async withModel<T>(slot: string, work: () => Promise<T>): Promise<T> {
    const lease = await this.acquire(slot);
    try {
      return await work();
    } finally {
      lease.release();
    }
  }

  private async acquire(slot: string): Promise<ModelLease> {
    if (this.activeLeases > 0 && !this.resident.includes(slot)) {
      // Something else is loaded and running: wait rather than risk the
      // out-of-memory error that would lose the whole run.
      await new Promise<void>((resolve) => this.queue.push({ slot, grant: resolve }));
    }

    this.activeLeases++;
    await this.makeResident(slot);

    let released = false;
    return {
      slot,
      release: () => {
        if (released) return;
        released = true;
        this.activeLeases--;
        this.drain();
      },
    };
  }

  private async makeResident(slot: string): Promise<void> {
    this.resident = this.resident.filter((s) => s !== slot);
    this.resident.push(slot);
    while (this.resident.length > this.concurrency) {
      const evicted = this.resident.shift();
      if (evicted) await this.onEvict?.(evicted);
    }
  }

  private drain(): void {
    if (this.activeLeases > 0) return;
    const next = this.queue.shift();
    next?.grant();
  }
}

File: services/perception/src/editorial_perception/scheduler.py
"""Keeping one large model in memory at a time.

Sixteen gigabytes is the declared target, and a transcriber, a vision encoder
and a vision-language model will not co-exist in it. The pipeline is already
staged so that everything is transcribed, then everything is embedded, then the
events that need it are described, which means the fix is not smaller models but
never holding two at once.

Loaders are lazy and cached, so "transcribe thirty files" costs one model load
rather than thirty, which is where nearly all of the wall-clock time goes.
"""

from __future__ import annotations

import gc
from collections.abc import Callable
from typing import Any


class ModelScheduler:
    def __init__(self, capacity: int = 1) -> None:
        self.capacity = max(1, capacity)
        self._loaded: dict[str, Any] = {}
        self._order: list[str] = []

    @property
    def resident(self) -> list[str]:
        return list(self._order)

    def get(self, slot: str, load: Callable[[], Any]) -> Any:
        """Returns the model for `slot`, loading it and evicting others as needed."""
        if slot in self._loaded:
            self._order.remove(slot)
            self._order.append(slot)
            return self._loaded[slot]

        while len(self._order) >= self.capacity:
            self.unload(self._order[0])

        model = load()
        self._loaded[slot] = model
        self._order.append(slot)
        return model

    def unload(self, slot: str) -> None:
        model = self._loaded.pop(slot, None)
        if slot in self._order:
            self._order.remove(slot)
        if model is None:
            return
        del model
        gc.collect()
        _empty_accelerator_cache()

    def unload_all(self) -> None:
        for slot in list(self._order):
            self.unload(slot)


def _empty_accelerator_cache() -> None:
    """Returns freed memory to the driver.

    Python's garbage collector releasing a model is not the same as the GPU
    getting its memory back, and without this the next load fails on a card that
    is, as far as the driver is concerned, still full.
    """
    try:
        import torch  # noqa: PLC0415
    except ImportError:
        return
    try:
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
    except Exception:  # noqa: BLE001 - freeing cache must never fail a run
        pass
